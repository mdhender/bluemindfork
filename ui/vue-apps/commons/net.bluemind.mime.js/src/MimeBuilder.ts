import Builder from "@bluemind/emailjs-mime-builder";
import { MessageBody } from "@bluemind/backend.mail.api";

export default class MimeBuilder {
    getContent: (part: MessageBody.Part) => Promise<Uint8Array | null>;

    constructor(getContent: (part: MessageBody.Part) => Promise<Uint8Array | null>) {
        this.getContent = getContent;
    }

    async build(part: MessageBody.Part): Promise<string | undefined> {
        return (await this.appendPart(part))?.build();
    }
    private async appendPart(part: MessageBody.Part): Promise<Builder | undefined> {
        if (part.mime) {
            const node = new Builder(part.mime);
            if (part.mime.startsWith("multipart/") && part.children) {
                for (const child of part.children) {
                    const newNode = await this.appendPart(child);
                    if (newNode) {
                        node.appendChild(newNode);
                    }
                }
            } else {
                const content = await this.getContent(part);
                if (content) {
                    node.setContent(content);
                }

                if (part.dispositionType) {
                    if (part.dispositionType === "ATTACHMENT" && part.fileName) {
                        node.setHeader("Content-Disposition", `attachment; filename="${part.fileName}`);
                    } else {
                        node.setHeader("Content-Disposition", part.dispositionType.toLowerCase());
                    }
                }
                if (part.contentId) {
                    node.setHeader("Content-ID", part.contentId);
                }
            }
            return node;
        }
        throw new Error("Malformed item");
    }
}
