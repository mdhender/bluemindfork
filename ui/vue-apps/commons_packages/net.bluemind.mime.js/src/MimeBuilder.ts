import Builder from "@bluemind/emailjs-mime-builder";
import { MessageBody } from "@bluemind/backend.mail.api";

export default class MimeBuilder {
    private getContent: (part: MessageBody.Part) => Promise<Uint8Array | string>;

    constructor(getContent: (part: MessageBody.Part) => Promise<Uint8Array | string>) {
        this.getContent = getContent;
    }

    async build(part: MessageBody.Part): Promise<string> {
        const node = await this.appendPart(part);
        return node.build();
    }

    private async appendPart(part: MessageBody.Part): Promise<Builder> {
        const isEncoded = !part.encoding; // if encoding is undefined, consider content is already encoded
        const node = new Builder(this.getContentType(part), { isEncoded });
        if (part.mime!.startsWith("multipart/") && part.children && part.children.length > 0) {
            for (const child of part.children) {
                const newNode = await this.appendPart(child);
                node.appendChild(newNode);
            }
        } else {
            const content = await this.getContent(part);
            node.setContent(content);
            if (part.dispositionType === "ATTACHMENT" && part.fileName) {
                node.setHeader("Content-Disposition", `attachment; filename="${part.fileName}`);
            } else if (part.dispositionType) {
                node.setHeader("Content-Disposition", part.dispositionType.toLowerCase());
            }
            if (part.contentId) {
                node.setHeader("Content-ID", part.contentId);
            }
        }
        part.headers?.forEach(header => {
            if (header.name && header.values && header.values[0]) {
                node.setHeader(header.name, header.values[0]);
            }
        });
        return node;
    }

    private getContentType(part: MessageBody.Part): string {
        const hasHeader = part.headers?.find(({ name }) => name?.toLowerCase() === "content-type");
        if (hasHeader) {
            return hasHeader.values![0];
        }
        return " " + part.mime + (part.charset ? "; charset=" + part.charset : "");
    }
}
