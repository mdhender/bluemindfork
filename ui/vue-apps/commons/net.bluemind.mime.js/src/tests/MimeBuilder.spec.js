import MimeBuilder from "../MimeBuilder";
import { Blob } from "buffer";
import basic from "./datas/bodyStructures/basic";
import alternative from "./datas/bodyStructures/alternative";
import withAttachment from "./datas/bodyStructures/withAttachment";
import withInlineImage from "./datas/bodyStructures/withInlineImage";

const getContent = async part => {
    let blob;
    if (part.mime === "text/plain") {
        blob = new Blob(["text"]);
    } else if (part.mime === "text/html") {
        blob = new Blob(["<div>text</div>"]);
    } else if (part.mime === "image/png") {
        blob = new Blob(["a mocked image"], { type: "image/png" });
    }
    const content = await blob.arrayBuffer();
    return Promise.resolve(new Uint8Array(content));
};
const builder = new MimeBuilder(getContent);

describe("MimeBuilder", () => {
    test("basic", async () => {
        const mimeTree = await builder.build(basic);
        expect(removeRandomBoundaries(mimeTree)).toMatchSnapshot();
    });
    test("alternative", async () => {
        const mimeTree = await builder.build(alternative);
        expect(removeRandomBoundaries(mimeTree)).toMatchSnapshot();
    });

    test("with attachment", async () => {
        const mimeTree = await builder.build(withAttachment);
        expect(removeRandomBoundaries(mimeTree)).toMatchSnapshot();
    });

    test("with inline image", async () => {
        const mimeTree = await builder.build(withInlineImage);
        expect(removeRandomBoundaries(mimeTree)).toMatchSnapshot();
    });
});

// Remove boundaries randomly generated to allow a snapshot match
function removeRandomBoundaries(mimeTree) {
    const regex = /---.*/g;
    return mimeTree.replaceAll(regex, "");
}
