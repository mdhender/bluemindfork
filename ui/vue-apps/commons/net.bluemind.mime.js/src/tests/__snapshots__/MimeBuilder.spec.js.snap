// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MimeBuilder alternative 1`] = `
"Content-Type: multipart/alternative;
 boundary=\\"


Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

text

Content-Type: text/html; charset=utf-8
Content-Transfer-Encoding: quoted-printable

<div>text</div>

"
`;

exports[`MimeBuilder basic 1`] = `
"Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

text"
`;

exports[`MimeBuilder with attachment 1`] = `
"Content-Type: multipart/mixed;
 boundary=\\"


Content-Type: multipart/alternative;
 boundary=\\"


Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

text

Content-Type: text/html; charset=utf-8
Content-Transfer-Encoding: quoted-printable

<div>text</div>



Content-Type: image/png; charset=us-ascii
Content-Disposition: attachment; filename=capture.png; size=79360
Content-Transfer-Encoding: base64

YSBtb2NrZWQgaW1hZ2U=

"
`;

exports[`MimeBuilder with inline image 1`] = `
"Content-Type: multipart/alternative;
 boundary=\\"


Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

text

Content-Type: multipart/related;
 boundary=\\"


Content-Type: text/html; charset=utf-8
Content-Transfer-Encoding: quoted-printable

<div>text</div>

Content-Type: image/png
Content-Disposition: inline
Content-Id: <37E0D401-EC3D-4A26-BB5E-0F02C2CB7413@bluemind.net>
Content-Transfer-Encoding: base64

YSBtb2NrZWQgaW1hZ2U=



"
`;
